python/cudf/cudf/core/join/_join_helpers.py

```diff
@@ -0,0 +1,206 @@
+from __future__ import annotations
+
+import warnings
+from typing import TYPE_CHECKING, Any, Iterable, Tuple
+
+import numpy as np
+import pandas as pd
+
+import cudf
+from cudf.core.dtypes import CategoricalDtype
+
+if TYPE_CHECKING:
+    from cudf.core.column import CategoricalColumn, ColumnBase
+    from cudf.core.frame import Frame
+
+
+class _Indexer:
+    # Indexer into a column (either a data column or index level).
+    #
+    # >>> df
+    #    a
+    # b
+    # 4  1
+    # 5  2
+    # 6  3
+    # >>> _Indexer("a", column=True).get(df)  # returns column "a" of df
+    # >>> _Indexer("b", index=True).get(df)  # returns index level "b" of df
+
+    def __init__(self, name: Any, column=False, index=False):
+        if column and index:
+            raise ValueError("Cannot specify both column and index")
+        self.name = name
+        self.column, self.index = column, index
+
+    def get(self, obj: Frame) -> ColumnBase:
+        # get the column from `obj`
+        if self.column:
+            return obj._data[self.name]
+        else:
+            if obj._index is not None:
+                return obj._index._data[self.name]
+        raise KeyError()
+
+    def set(self, obj: Frame, value: ColumnBase, validate=False):
+        # set the colum in `obj`
+        if self.column:
+            obj._data.set_by_label(self.name, value, validate=validate)
+        else:
+            if obj._index is not None:
+                obj._index._data.set_by_label(
+                    self.name, value, validate=validate
+                )
+            else:
+                raise KeyError()
+
+
+def _frame_select_by_indexers(
+    frame: Frame, indexers: Iterable[_Indexer]
+) -> Frame:
+    # Select columns from the given `Frame` using `indexers`,
+    # and return a new `Frame`.
+    index_data = frame._data.__class__()
+    data = frame._data.__class__()
+
+    for idx in indexers:
+        if idx.index:
+            index_data.set_by_label(idx.name, idx.get(frame), validate=False)
+        else:
+            data.set_by_label(idx.name, idx.get(frame), validate=False)
+
+    result_index = cudf.Index._from_data(index_data) if index_data else None
+    result = cudf.core.frame.Frame(data=data, index=result_index)
+    return result
+
+
+def _match_join_keys(
+    lcol: ColumnBase, rcol: ColumnBase, how: str
+) -> Tuple[ColumnBase, ColumnBase]:
+    # returns the common dtype that lcol and rcol should be casted to,
+    # before they can be used as left and right join keys.
+    # If no casting is necessary, returns None
+
+    common_type = None
+
+    # cast the keys lcol and rcol to a common dtype
+    ltype = lcol.dtype
+    rtype = rcol.dtype
+
+    # if either side is categorical, different logic
+    if isinstance(ltype, CategoricalDtype) or isinstance(
+        rtype, CategoricalDtype
+    ):
+        return _match_categorical_dtypes(lcol, rcol, how)
+
+    if pd.api.types.is_dtype_equal(ltype, rtype):
+        return lcol, rcol
```

Comment:
Don't need to solve it in this PR, but does this handle Decimal dtypes? If yes then we likely need additional decimal typecasting logic below this, if no then we need to handle Decimal equality checking as well.

Can be in a follow up PR.

Reply:
I think this is what https://github.com/rapidsai/cudf/pull/7670/ is tackling. WIll fix there.